pkgs/development/compilers/gcc/common/configure-flags.nix

```
@@ -114,6 +114,14 @@ let
     ++ lib.optional (!(crossMingw && crossStageStatic))
       "--with-native-system-header-dir=${lib.getDev stdenv.cc.libc}/include"
 
+    # cross-built gcc needs to be tricked as if it is built with headers
+    # available in sysroot. For that we need to pass some --with-sysroot= value:
+    #   https://github.com/NixOS/nixpkgs/pull/181802#issuecomment-1186822355
+    # Without this flag --with-native-system-header-dir= has no effect
+    # and target libc features are not detected like stack protector support
+    ++ lib.optional (buildPlatform != hostPlatform && hostPlatform == targetPlatform)
+      "--with-sysroot=/"
```

```suggestion
      "--with-build-sysroot=/"
```

Could we use this instead?  I think [it would allow us to remove the `hostPlatform==targetPlatform` check](https://gcc.gnu.org/git/?p=gcc.git;a=blob;f=gcc/configure.ac;h=446747311a6aec3c810ad6aa4190f7bd383b94f7;hb=HEAD#l2439) above; then we don't have to explain why that case is excluded.  Also, it should work for canadian cross-compilers (although I don't think nixpkgs supports them right now, but it could someday).

From the description of the `--with-build-sysroot` flag it sounds closer to what we want: it's telling build-gcc where to look for the target system's sysroot, without influencing where host-gcc will look.